File: modules/sql/app/model/querycorrectors/create/CreateCorrector.java
package model.querycorrectors.create;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import model.exercise.SqlExercise;
import model.matching.Match;
import model.matching.MatchingResult;
import model.querycorrectors.ColumnWrapper;
import model.querycorrectors.QueryCorrector;
import model.querycorrectors.SqlExecutionResult;
import net.sf.jsqlparser.expression.Expression;
import net.sf.jsqlparser.schema.Table;
import net.sf.jsqlparser.statement.create.table.CreateTable;
import play.db.Database;

public class CreateCorrector extends QueryCorrector<CreateTable> {
  
  public CreateCorrector() {
    super("CREATE TABLE");
  }

  @Override
  protected SqlExecutionResult executeQuery(Database database, CreateTable userStatement, CreateTable sampleStatement,
      SqlExercise exercise) {
    return null;
  }
  
  @Override
  protected List<ColumnWrapper> getColumnWrappers(CreateTable query) {
    return query.getColumnDefinitions().parallelStream().map(ColumnWrapper::wrap).collect(Collectors.toList());
  }
  
  @Override
  protected List<String> getTableNames(CreateTable query) {
    return Arrays.asList(query.getTable().toString());
  }
  
  @Override
  protected List<Table> getTables(CreateTable query) {
    return Arrays.asList(query.getTable());
  }
  
  @Override
  protected Expression getWhere(CreateTable query) {
    return null;
  }
  
  @Override
  protected List<MatchingResult<? extends Object, ? extends Match<? extends Object>>> makeOtherComparisons(
      CreateTable userQ, CreateTable sampleQ) {
    return Collections.emptyList();
  }
  
}

File: app/model/sql/CreateCorrector.scala
package model.sql

import model.core.matching.{Match, MatchingResult}
import net.sf.jsqlparser.expression.Expression
import net.sf.jsqlparser.schema.Table
import net.sf.jsqlparser.statement.Statement
import net.sf.jsqlparser.statement.create.table.CreateTable

import scala.collection.JavaConverters.asScalaBufferConverter
import scala.util.{Failure, Success, Try}

object CreateCorrector extends QueryCorrector("CREATE TABLE") {

  override type Q = CreateTable

  override protected def getColumnWrappers(query: Q): Seq[CreateColumnWrapper] = query.getColumnDefinitions.asScala map ColumnWrapper.wrapColumn

  override protected def getTables(query: Q): Seq[Table] = Seq(query.getTable)

  override protected def getWhere(query: Q): Option[Expression] = None

  override protected def checkStatement(statement: Statement): Try[CreateTable] = statement match {
    case q: CreateTable   => Success(q)
    case other: Statement => Failure(WrongStatementTypeException(queryType, gotten = other.getClass.getSimpleName))
  }

  override protected def performAdditionalComparisons(userQuery: CreateTable, sampleQuery: CreateTable): Seq[MatchingResult[_, _ <: Match[_]]] = Seq.empty

}
